### DISCUSSION TITLE
Why a postprocessed variable displays zero?

### DISCUSSION THREAD

**Discussion by amerallaf** [idaholab/moose#27786] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hello,

I am trying to postprocess the heat_source variable (source_integral), but it keeps showing me the value as zero. The heat_source has a greater value than zero, because whenever I change it, the maximum temperature value also changes, but I still get the source_integral as zero.

I would appreciate your help in solving this issue.

Here is my input:

```c
L = 300.0
T_fluid = ${fparse 280.0 + 273.15}

[Mesh]
  [file]
    type = FileMeshGenerator
    file = mesh_in.e
  []
[]

[Variables]
  [temp]
    initial_condition = ${T_fluid}
  []
[]

[AuxVariables]
  [./heat_source]

  [../]
[]

[Kernels]
  [hc]
    type = HeatConduction
    variable = temp
  []
  [heat_source]
    type = HeatSource
    value = 250 
    variable = temp
    block = '2 3'
  []
[]

[Functions]
  [T_fluid]
    type = ParsedFunction
    expression = '573.0 + 50.0 * (z / 300.0)'
  []
[]

[BCs]
  [surface]
    type = ConvectiveFluxFunction
    T_infinity = T_fluid

    # convert from W/m2/K to W/cm2/K
    coefficient = ${fparse 1000.0/100.0/100.0}
    variable = temp
    boundary = 'rmax_c'
  []
[]

[ThermalContact]
  # This adds boundary conditions bewteen the fuel and the cladding, which represents
  # the heat flux in both directions as
  # q''= h * (T_1 - T_2)
  # where h is a conductance that accounts for conduction through a material and
  # radiation between two infinite parallel plate gray bodies.
  [one_to_two]
    type = GapHeatTransfer
    variable = temp
    primary = 'rmax'
    secondary = 'rmin_c'

    # we will use a quadrature-based approach to find the gap width and cross-side temperature
    quadrature = true

    # emissivity of the fuel
    emissivity_primary = 0.8

    # emissivity of the clad
    emissivity_secondary = 0.8

    # thermal conductivity of the gap material
    gap_conductivity = 1.0

    # geometric terms related to the gap
    gap_geometry_type = CYLINDER
    cylinder_axis_point_1 = '0 0 0'
    cylinder_axis_point_2 = '0 0 ${L}'
  []
[]

[Materials]
  [k_clad]
    type = GenericConstantMaterial
    prop_values = '0.5'
    prop_names = 'thermal_conductivity'
    block = '1'
  []
  [k_fuel]
    type = GenericConstantMaterial
    prop_values = '0.05'
    prop_names = 'thermal_conductivity'
    block = '2 3'
  []
[]

[Executioner]
  type = Transient
  nl_abs_tol = 1e-8
  num_steps = 5
  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'
[]

[Outputs]
  exodus = true
  print_linear_residuals = false
[]

[Postprocessors]
  [source_integral]
    type = ElementIntegralVariablePostprocessor
    variable = heat_source
  []
  [max_T]
    type = NodalExtremeValue
    variable = temp
  []
[]
``` 
and this is the mesh input

```c
r_pin = 0.39218
clad_ir = 0.40005
clad_or = 0.45720
L = 300.0

[Mesh]
  [clad] # This makes a circular annulus that will represent the clad
    type = AnnularMeshGenerator
    nr = 3
    nt = 20
    rmin = ${clad_ir}
    rmax = ${clad_or}
    quad_subdomain_id = 1
    tri_subdomain_id = 0
  []
  [extrude_clad] # this extrudes the circular annulus in the axial direction
    type = AdvancedExtruderGenerator
    input = clad
    heights = '${L}'
    num_layers = '40'
    direction = '0 0 1'
  []

  # A sideset in MOOSE is both an ID and a name.
  # This renames the sideset numbers on the clad inner surface (0) and outer surface (1)
  # to (5) and (4), respectively. This is done to avoid name collisions with another
  # AnnularMeshGenerator we use for the fuel pellet. After we rename the sideset IDs, we
  # need another RenameBoundaryGenerator to rename the sideset names.
  [rename_clad]
    type = RenameBoundaryGenerator
    input = extrude_clad
    old_boundary = '1 0' # outer surface, inner surface
    new_boundary = '5 4'
  []
  [rename_clad_names]
    type = RenameBoundaryGenerator
    input = rename_clad
    old_boundary = 'rmax rmin' # outer surface, inner surface
    new_boundary = 'rmax_c rmin_c'
  []

  [fuel] # this makes a circle that will represent the fuel
    type = AnnularMeshGenerator
    nr = 10
    nt = 20
    rmin = 0
    rmax = ${r_pin}
    quad_subdomain_id = 2
    tri_subdomain_id = 3
    growth_r = -1.2
  []
  [extrude] # this extrudes the circle in the axial direction
    type = AdvancedExtruderGenerator
    input = fuel
    heights = '${L}'
    num_layers = '40'
    direction = '0 0 1'
  []
  [combine]
    type = CombinerGenerator
    inputs = 'rename_clad_names extrude'
  []

  # one of the mesh generators does not work with distributed mesh
  parallel_type = replicated
[]

```

**Comment by GiudGiud** [comment]
Hello

In paraview are you visualizing the last time step?

The default is to show you the initial condition

**Reply by amerallaf** [reply]
Hello @GiudGiud,
I am checking all time steps and it is zero always.

See the figure below:
![Capture2](https://github.com/idaholab/moose/assets/155470525/5435b7ca-599e-4399-91d5-2a0ddf9646a3)
But the interesting thing is if I change the power the temperature changes accordingly:

For heat_source=250

![Capture1](https://github.com/idaholab/moose/assets/155470525/9cb8f8cc-6154-43ea-9592-c060fb347cd2)

For heat_source=350

![Capture1_350](https://github.com/idaholab/moose/assets/155470525/7fc2d688-8ef4-4109-b861-c6d503484c72)

**Reply by GiudGiud** [reply]
Can you look at the field instead?

This graph is likely showing a value at the wrong location

**Reply by GiudGiud** [reply]
Maybe not

How did you compute heat_source in the simulation ?

**Reply by GiudGiud** [reply]
In your input the postprocessor is called source_integral not heat_source. So what are you plotting in that graph?

**Reply by amerallaf** [reply]
Sorry for the confusion, it is the same variable, I just ran a different version of the script

```c
  [heat_source]
    type = ElementIntegralVariablePostprocessor
    variable = heat_source
  []
``` 

I also tried the field and still shows zero (for any time step).

**Reply by GiudGiud** [reply]
I think there is some confusion here
you are creating a heat source in the equation with this kernel
```
  [heat_source]
    type = HeatSource
    value = 250 
    variable = temp
    block = '2 3'
  []
```
which does NOT use a heat_source auxiliary variable. That auxiliary variable is currently doing nothing, sitting at a 0 value.

If you want to use this variable:
- use an auxiliary kernel or a transfer to set the heat source aux variable
- use a CoupledForce kernel to make use of the heat_source variable in the equations

**Reply by amerallaf** [reply, marked as answer]
Thanks a lot @GiudGiud , now it is working well.

I did the following in the end:

```c
[AuxVariables]
  [power]
    block = '2 3'
    initial_condition = 300
  []
[]

[Kernels]
  [hc]
    type = HeatConduction
    variable = temp
  []
  [source]
    type = CoupledForce
    variable = temp
    v = power
    block = '2 3'
  []
[]
```

**Reply by amerallaf** [reply]
Currently I am trying to turn this variable into a function of time by doing the following:

```c
[AuxVariables]
  [power]
    block = '2 3'
    initial_condition = 300
  []
[]

[AuxKernels]
  [./reconstruct_power]
    type = FunctionAux
    variable = power
    function = Fiss_Function
  [../]
[]

[Kernels]
  [source]
    type = CoupledForce
    variable = temp
    v = power
    block = '2 3'
  [../]
[]

[Functions]
  [Fiss_Function]
    type = ParsedFunction
    expression = 't* ${power}'
  []
[]
``` 

But I am getting the following error: 

*** ERROR ***
no variable 'power' found for substitution expression

I would really appreciate your advice on this.

**Reply by GiudGiud** [reply]
Nothing in the snippet you pasted asks for a source auxiliary variable?

**Reply by amerallaf** [reply]
No this is the whole thing:

*** ERROR ***
..../solid_CoupledForce_time_1.i:54.5: no variable 'power' found for substitution expression


Stack frames: 11
0: libMesh::print_trace(std::ostream&)
1: moose::internal::mooseErrorRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
2: void mooseError<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)
3: Moose::Builder::build()
4: MooseApp::setupOptions()
5: MooseApp::run()
6: ./cardinal-opt(+0x3264) [0x7f5694ed0264]
7: main
8: /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7f5685af9d90]
9: __libc_start_main
10: ./cardinal-opt(+0x34dd) [0x7f5694ed04dd]
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
:
system msg for write_line failure : Bad file descriptor

**Reply by GiudGiud** [reply]
Oh this is from your ParsedFunction

The power variable in that expression must be defined in the input file as a simple 

Power = some_value

**Reply by amerallaf** [reply]
Sorry I am confused a bit, what I am trying to do is turning the AuxVariable:power into a function(time), that's why I presented this expression using type = ParsedFunction, if I make it as "simple", then how do I define it as a function?

**Reply by GiudGiud** [reply]
a field variable cannot go in a function
A function can only depend on space and time

Use a material property (ParsedMaterial) if you want to define a function of the power field variable

**Reply by amerallaf** [reply]
But in my practice, I am working based on the following example:
**(modules/fsi/test/tests/2d-finite-strain-steady/thermal-me.i)**
In this example, a heat_source variable is defined and reconstructed based on a function of y-axis.

I reworked the whole thing to follow the mentioned example, my latest input is as follows:

```c
L = 300.0
T_fluid = 550.0

[Mesh]
  [file]
    type = FileMeshGenerator
    file = mesh_in.e
  []
[]

[Variables]
  [temp]
    initial_condition = ${T_fluid}
  []
[]

[AuxVariables]
  [heat_source]
    family = MONOMIAL
    order = FIRST
    block = '2 3'
  []
[]

[Kernels]
  [heat_source]
    type = ADCoupledForce
    variable = temp
    v = heat_source
    block = '2 3'
    use_displaced_mesh = true
  []
[]

[AuxKernels]
  [heat_source_function_auxk]
    type = FunctionAux
    variable = heat_source
    function = heat_source_function
    block = '2 3'
  []
[]

[Functions]
  [heat_source_function]
    type = ParsedFunction
    expression = '100 * t'
  []
[]

[BCs]
  [surface]
    type = ConvectiveFluxFunction
    T_infinity = T_fluid

    # convert from W/m2/K to W/cm2/K
    coefficient = ${fparse 1000.0/100.0/100.0}
    variable = temp
    boundary = 'rmax_c'
  []
[]

[ThermalContact]
  # This adds boundary conditions bewteen the fuel and the cladding, which represents
  # the heat flux in both directions as
  # q''= h * (T_1 - T_2)
  # where h is a conductance that accounts for conduction through a material and
  # radiation between two infinite parallel plate gray bodies.
  [one_to_two]
    type = GapHeatTransfer
    variable = temp
    primary = 'rmax'
    secondary = 'rmin_c'

    # we will use a quadrature-based approach to find the gap width and cross-side temperature
    quadrature = true

    # emissivity of the fuel
    emissivity_primary = 0.8

    # emissivity of the clad
    emissivity_secondary = 0.8

    # thermal conductivity of the gap material
    gap_conductivity = 1.0

    # geometric terms related to the gap
    gap_geometry_type = CYLINDER
    cylinder_axis_point_1 = '0 0 0'
    cylinder_axis_point_2 = '0 0 ${L}'
  []
[]

[Materials]
  [k_clad]
    type = GenericConstantMaterial
    prop_values = '0.5'
    prop_names = 'thermal_conductivity'
    block = '1'
  []
  [k_fuel]
    type = GenericConstantMaterial
    prop_values = '0.05'
    prop_names = 'thermal_conductivity'
    block = '2 3'
  []
[]

[Executioner]
  type = Transient
  nl_abs_tol = 1e-8
  num_steps = 5
  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'
[]

[Outputs]
  exodus = true
  print_linear_residuals = false
[]

[Postprocessors]
  [source_inegrated]
    type = ElementIntegralVariablePostprocessor
    variable = heat_source
    block = '2 3'
  []
  [max_T]
    type = NodalExtremeValue
    variable = temp
  []
[]
``` 

Now I am getting a different error:

*** ERROR ***
The following error occurred in the Problem 'MOOSE Problem' of type FEProblem.

Unable to find function T_fluid

**Reply by GiudGiud** [reply]
You need to create a T_fluid function. An object is requesting it. Dont you get more context on this error? Like the object that requests this?

**Reply by GiudGiud** [reply]
Basically objects in MOOSE expect inputs of a certain type. If you do not have the input in the desired type you either:
- use a different object for the needed purpose
- create a new object that uses the input you have (for example, create the 'variable' version of the object that expects a function
- create new input objects. For example a material property can be created from a function or from a variable with the GenericFunctionMaterial or the ParsedMaterial

**Reply by amerallaf** [reply]
Sorry, it is a bit hard for me to understand where to find which object is requesting it, but this is the whole error if it can be helpful:
```
*** ERROR ***
The following error occurred in the Problem 'MOOSE Problem' of type FEProblem.

Unable to find function T_fluid

Stack frames: 21
0: libMesh::print_trace(std::ostream&)
1: moose::internal::mooseErrorRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
2: MooseBase::callMooseError(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool) const
3: void MooseBaseErrorInterface::mooseError<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&&) const
4: FEProblemBase::getFunction(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned int)
5: ConvectiveFluxFunction::ConvectiveFluxFunction(InputParameters const&)
6: RegistryEntry<ConvectiveFluxFunction>::build(InputParameters const&)
7: Factory::createUnique(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, InputParameters const&, unsigned int, bool)
8: std::unique_ptr<BoundaryCondition, std::default_delete<BoundaryCondition> > Factory::createUnique<BoundaryCondition>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, InputParameters const&, unsigned int)
9: std::shared_ptr<BoundaryCondition> Factory::create<BoundaryCondition>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, InputParameters const&, unsigned int)
10: NonlinearSystemBase::addBoundaryCondition(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, InputParameters&)
11: Action::timedAct()
12: ActionWarehouse::executeActionsWithAction(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
13: ActionWarehouse::executeAllActions()
14: MooseApp::runInputFile()
15: MooseApp::run()
16: ./cardinal-opt(+0x3264) [0x7f6be95cc264]
17: main
18: /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7f6bda1f9d90]
19: __libc_start_main
20: ./cardinal-opt(+0x34dd) [0x7f6be95cc4dd]
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
:
system msg for write_line failure : Bad file descriptor 
```

**Reply by GiudGiud** [reply]
5: ConvectiveFluxFunction::ConvectiveFluxFunction(InputParameters const&)

convectiveFluxFunction is expecting a function for T_fluid
This is your boundary condition.

What do you want to use for T_fluid? A variable? Or do you have a fixed profile (that is already known, that you do not need to solve for) in which case a function is a good way to represent T_fluid

**Reply by amerallaf** [reply]
Oh I see, thank you so much for your explaination.

Yes, a variable. I don't have a fixed/known profile.

**Reply by GiudGiud** [reply]
Maybe look for another BC that uses T_fluid as a variable then
